docs/r3/C5-Ethereum-Interop-Architecture.md

```diff
@@ -0,0 +1,379 @@
+
+## How do we communicate with an EVM from a flow?
+
+Now that we have a structure for encoding our functions, how do we send our encoded function out to the EVM?
+
+![](img/ExtInterop_C5ARB3.png)
+
+Well… we can’t.  We’re sandboxed.  In Corda 5, all flows are isolated from the outside world and access is restricted.  This is a security measure and one we certainly want to keep.
+
+Luckily, Corda 5 has a mechanism for getting operations done outside of a flow.
+
+
+## External Events
+
+In Corda 5, the mechanism for sending a message outside of the flow is called `External Events`.  This mechanism suspends the flow and sends the message out to the message bus.  It can then be picked up by a separate (non-sandboxed) processor which can convert it to a command to send to the EVM network.
+
+That processor will then return any response via the bus back to the flow, which will then resume.
+
+![](img/ExtInterop_C5ARB4.png)
+
+
+## What do we do when Corda Identity is not the same as an Ethereum wallet address?
+
+Corda and Ethereum have different concepts of what it means to have an "identity".  In Corda, the identity is bound to a _real world entity_, whereas, in Ethereum, it's a wallet address (so, not really an identity at all)
+
+What do we do when Corda’s  __Alice__  identity isn’t an exact match for her Ethereum wallet address?
+
+|  |  |
+|---|---|
+| Corda Identity: | <span style="color:#6A8759">“O = Alice, L = London, C = GB"</span> | |  |  |  |
+| Wallet Address: | <span style="color:#0070C0">0x02643ed6819E63d022434b5a6de6E01C73Efe571</span> |  |  |  |
+
+
+We can use the Alias mappings from C5-C5 Interop to go from a Corda Identity -> Ethereum wallet address.
+
+| | |
+| :-: | :- |
+| Interop Alias | An Interop alias is an alias or label assigned to an ego/alter ego to be used in an Interop group.|
+
+
+
+
+## Transactions need to be signed
+
+Ethereum requires us to signing a transaction before submitting it.  And we'll need to sign with a key that is known by the EVM as one belonging to a specific wallet.  
+
+Unfortunately, the current Crypto Worker doesn’t support external (to itself) private keys.  
+
+![](img/ExtInterop_C5ARB5.png)
+
+We will need to add to the Crypto Worker the operations we need:
+* Customer needed operations:
+  * Signing using existing keys
+    * Bring your own keys (BYOK)
+    * allowing us to import keys that are already known by the EVM
+  * External Signing services
+    * allowing us to have a transaction signed by some external service (e.g. HSM, or cloud provider)
+
+![](img/ExtInterop_C5ARB6.png)
+
+## When does a flow resume from a call?
+
+On Ethereum there are, broadly, two types of calls we can make:
+  * Queries – where a result is returned immediately
+  * Transactions – where a transaction hash is returned which can be queried to determine the result/receipt of a transaction

```

Comment:
this is a bit circular.  How about: "Transactions - which are signed instructions to perform some action on the underlying EVM network. These instructions may or may not be valid, and may or may not be executed and confirmed in any given timescale, so an ID (transaction hash) is returned, which can be queried to determine the result.